gensim/models/nmf.py

```diff
@@ -564,36 +528,90 @@ def _setup(self, corpus):
         self.A = np.zeros((self.num_topics, self.num_topics))
         self.B = np.zeros((self.num_tokens, self.num_topics))
 
-    def update(self, corpus):
+    def l2_norm(self, v):
+        Wt = self._W.T
+
+        l2 = 0
+
+        for doc, doc_topics in zip(v.T, self._h.T):
+            l2 += np.sum(np.square((doc - doc_topics.dot(Wt))))
+
+        return np.sqrt(l2)
+
+    def update(self, corpus, chunksize=None, passes=None, eval_every=None):
         """Train the model with new documents.
 
         Parameters
         ----------
-        corpus : iterable of list of (int, float), optional
+        corpus : iterable of list of (int, float) or `csc_matrix` with the shape (n_tokens, n_documents)
             Training corpus.
             Can be either iterable of documents, which are lists of `(word_id, word_count)`,
             or a sparse csc matrix of BOWs for each document.
             If not specified, the model is left uninitialized (presumably, to be trained later with `self.train()`).
+        chunksize: int, optional
+            Number of documents to be used in each training chunk.
+        passes: int, optional
+            Number of full passes over the training corpus.
+            Leave at default `passes=1` if your input is an iterator.
+        eval_every: int, optional
+            Number of batches after which l2 norm of (v - Wh) is computed. Decreases performance if set too low.
 
         """
-        if self._W is None:
-            self._setup(corpus)
 
-        chunk_idx = 1
+        # use parameters given in constructor, unless user explicitly overrode them
+        if passes is None:
+            passes = self.passes
+        if eval_every is None:
+            eval_every = self.eval_every
+
+        lencorpus = np.inf
+
+        if isinstance(corpus, scipy.sparse.csc.csc_matrix):
+            lencorpus = corpus.shape[1]
+        else:
+            try:
+                lencorpus = len(corpus)
+            except TypeError:
+                logger.info("input corpus stream has no len()")
+
+        if chunksize is None:
+            chunksize = min(lencorpus, self.chunksize)
+
+        evalafter = min(lencorpus, (eval_every or 0) * chunksize)
+
+        if lencorpus == 0:
+            logger.warning("Nmf.update() called with an empty corpus")
+            return
 
-        for _ in range(self.passes):
+        if isinstance(corpus, collections.Iterator) and self.passes > 1:
+            raise ValueError("Corpus is an iterator, only `passes=1` is valid.")
+
+        logger.info(
+            "running NMF training, %s topics, %i passes over the supplied corpus of %i documents, evaluating l2 norm "
+            "every %i documents",
+            self.num_topics, passes, lencorpus if lencorpus < np.inf else "?", evalafter,
```

Comment:
The formatstr expects an integer, but you will end up passing a string if lencorpus is np.inf.